src/lua/protobuf.lua

```diff
@@ -0,0 +1,642 @@
+                break
+            end
+        end
+        if not found then
+            error(('Field %q of %q type gets %q type value.'):format(
+                field_def.name, field_def.type, type(value)))
+        end
+        return
+    end
+    assert(type(exp_type) == 'string')
+    if type(value) ~= exp_type then
+        error(('Field %q of %q type gets %q type value.'):format(
+            field_def.name, field_def.type, type(value)))
+    end
+    return
+end
+
+local function validate_range(field_def, value, range)
+    local min = range ~= nil and range[1] or nil
+    local max = range ~= nil and range[2] or nil
+    -- To avoid incorrect comparison unsigned cdata should not be
+    -- checked for lower limit. If one of the limits is 'nil' this
+    -- function skips the comparison with this limit.
+    if min ~= nil and value < min or max ~= nil and value > max then
+        error(('Input data for %q field is %q and do not fit in %q')
+            :format(field_def.name, value, field_def.type))
+    end
+end
+
+validate_scalar = function(field_def, value)
+    local scalar_def = scalars[field_def.type]
+    local value_type = type(value)
+    assert(scalar_def.accept_type ~= nil)
+    -- Checks type of input according to the allowed types for this field.
+    validate_type(field_def, value, scalar_def.accept_type)
+    -- Checks length of the string if input type assumed to be string.
+    if scalar_def.accept_type == 'string' then
+        validate_length(value)
+    end
+    -- Checks number values for being NaN or inf.
+    if value_type == 'number' and is_nan(value) then
+        error(('Input data for %q field is NaN'):format(field_def.name))
+    end
+    if value_type == 'number' and is_inf(value) then
+        error(('Input data for %q field is inf'):format(field_def.name))
+    end
+    -- Checks values assumed to be integer.
+    if scalar_def.integral_only then
+        check_integer(field_def, value)
+    end
+    -- Checks numeric values to see if they belong to the limits.
+    if scalar_def.limits ~= nil then
+        if value_type == 'cdata' then
+            value_type = ffi.istype(int64_t, value) and 'int64'
+                or 'uint64'
+        end
+        validate_range(field_def, value, scalar_def.limits[value_type])
+    end
+end
+
+-- }}} Validation
+
+
+-- {{{ Encoders
+
+local function encode_repeated(protocol, field_def, value)
+    local buf = {}
+    local encode_as_packed = false
+    if type(value) ~= 'table' then
+        error('For repeated fields table data are needed')
+    end
+    validate_table_is_array(field_def, value)
+    if is_scalar(field_def) then
+        local scalar_def = scalars[field_def.type]
+        encode_as_packed = scalar_def.encode_as_packed
+    end
+    for _, item in ipairs(value) do
+        local encoded_item = encode_field(protocol, field_def, item, true)
+        if encode_as_packed then
+            encoded_item = string.sub(encoded_item, 2)
+        end
```

Comment:
I agree it's my mistake. Added function remove_tag() which deletes tag from encoded repeated field

Reply:
This solution seems OK. A small nit:

```lua
return string.sub(value, tag_length)
```

The last argument is a length of the tag plus 1. The mplementation is correct, but it stores tag length plus 1 in the `tag_length` variable that looks a bit confusing.

IOW, let's start from `local tag_length = 0` instead of 1.

Out of curiosity: is this function actually takes `value == ''` on some of our test cases?

Reply:
Ok, I started from `tag_length = 0` but it will be `string.sub(value, tag_length + 1)` because second argument of `string.sub()` is number of first symbol for new string.

Reply:
`string.sub(value, tag_length + 1)` is completely OK: the first symbol of the new string starts on the tag length plus 1 in the old string. And now it is what actually written in the code.

Regarding the `value == ''` check: I'm not sure that we should just skip missed/null/default array values. Thanks for pointing the test case. I continued this discussion in https://github.com/tarantool/tarantool/pull/9488#discussion_r1676519994.